lib/srv/alpnproxy/aws_local_proxy.go

```diff
@@ -46,11 +56,168 @@ func (m *AWSAccessMiddleware) CheckAndSetDefaults() error {
+	if err != nil {
+		m.Log.WithError(err).Error("Failed to parse AWS request authorization header.")
+		rw.WriteHeader(http.StatusForbidden)
+		return true
+	}
+
+	if assumedRole, found := m.findAssumedRole(sigV4.KeyID); found {
+		return m.handleRequestByAssumedRole(rw, req, assumedRole)
+	}
+	return m.handleCommonRequest(rw, req)
+}
+
+func (m *AWSAccessMiddleware) handleCommonRequest(rw http.ResponseWriter, req *http.Request) bool {
+	if err := awsutils.VerifyAWSSignature(req, m.AWSCredentials); err != nil {
 		m.Log.WithError(err).Error("AWS signature verification failed.")
 		rw.WriteHeader(http.StatusForbidden)
 		return true
 	}
 	return false
 }
+
+func (m *AWSAccessMiddleware) handleRequestByAssumedRole(rw http.ResponseWriter, req *http.Request, assumedRole *sts.AssumeRoleOutput) bool {
+	credentials := credentials.NewStaticCredentials(
+		aws.ToString(assumedRole.Credentials.AccessKeyId),
+		aws.ToString(assumedRole.Credentials.SecretAccessKey),
+		aws.ToString(assumedRole.Credentials.SessionToken),
+	)
+
+	if err := awsutils.VerifyAWSSignature(req, credentials); err != nil {
+		m.Log.WithError(err).Error("AWS signature verification failed.")
+		rw.WriteHeader(http.StatusForbidden)
+		return true
+	}
+
+	m.Log.Debugf("Rewriting headers for AWS request by assumed role %q.", aws.ToString(assumedRole.AssumedRoleUser.Arn))
+
+	// Add a custom header for marking the special request.
+	req.Header.Add(appcommon.TeleportAWSAssumedRole, aws.ToString(assumedRole.AssumedRoleUser.Arn))
+
+	// Rename the original authorization header to ensure older app agents
+	// (that don't support the requests by assumed roles) will fail.
+	utils.RenameHeader(req.Header, awsutils.AuthorizationHeader, appcommon.TeleportAWSAssumedRoleAuthorization)
+	return false
+}
+
+func (m *AWSAccessMiddleware) HandleResponse(response *http.Response) error {
+	if response == nil || response.Request == nil {
+		return nil
+	}
+
+	authHeader := utils.GetAnyHeader(response.Request.Header,
+		awsutils.AuthorizationHeader,
+		appcommon.TeleportAWSAssumedRoleAuthorization,
+	)
+
+	sigV4, err := awsutils.ParseSigV4(authHeader)
+	if err != nil {
+		m.Log.WithError(err).Error("Failed to parse AWS request authorization header.")
+		return nil
+	}
+
+	if strings.EqualFold(sigV4.Service, sts.ServiceID) {
+		return trace.Wrap(m.handleSTSResponse(response))
+	}
+	return nil
+}
+
+func (m *AWSAccessMiddleware) handleSTSResponse(response *http.Response) error {
+	// Only looking for successful sts:AssumeRole calls.
+	if response.Request.Method != http.MethodPost ||
+		response.StatusCode != http.StatusOK {
+		return nil
+	}
+
+	// In case something goes wrong when draining the body, return an error.
+	body, err := utils.GetAndReplaceResponseBody(response)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Save the credentials if valid AssumeRoleOutput is found.
+	type AssumeRoleResponse struct {
+		AssumeRoleResult sts.AssumeRoleOutput
+	}
```

Comment:
is the type def needed?

Reply:
If you check the sample XML response, the response XML consists of `AssumeRoleResult` and `ResponseMetadata`. I've added `xml:AssumeRoleResult` tag and hope that's a little cleaer.